RecoLocalCalo/EcalRecProducers/plugins/alpaka/EcalRecHitProducerPortable.cc

```diff
@@ -0,0 +1,179 @@
+    // output
+    using OutputProduct = EcalRecHitDeviceCollection;
+    const device::EDPutToken<OutputProduct> recHitsTokenEB_;
+    device::EDPutToken<OutputProduct> recHitsTokenEE_;
+
+    // configuration parameters
+    ecal::rechit::ConfigurationParameters configParameters_;
+
+    // conditions tokens
+    const device::ESGetToken<EcalRecHitConditionsDevice, EcalRecHitConditionsRcd> recHitConditionsToken_;
+    const device::ESGetToken<EcalRecHitParametersDevice, JobConfigurationGPURecord> recHitParametersToken_;
+  };
+
+  void EcalRecHitProducerPortable::fillDescriptions(edm::ConfigurationDescriptions& confDesc) {
+    edm::ParameterSetDescription desc;
+
+    desc.add<edm::InputTag>("uncalibrecHitsInLabelEB",
+                            edm::InputTag("ecalMultiFitUncalibRecHitPortable", "EcalUncalibRecHitsEB"));
+    desc.add<std::string>("recHitsLabelEB", "EcalRecHitsEB");
+    desc.add<bool>("killDeadChannels", true);
+    desc.add<bool>("recoverEBIsolatedChannels", false);
+    desc.add<bool>("recoverEBVFE", false);
+    desc.add<bool>("recoverEBFE", true);
+
+    desc.add<double>("EBLaserMIN", 0.5);
+    desc.add<double>("EBLaserMAX", 3.0);
+
+    desc.ifValue(edm::ParameterDescription<bool>("isPhase2", false, true),
+                 false >> (edm::ParameterDescription<edm::InputTag>(
+                               "uncalibrecHitsInLabelEE",
+                               edm::InputTag("ecalMultiFitUncalibRecHitPortable", "EcalUncalibRecHitsEE"),
+                               true) and
+                           edm::ParameterDescription<std::string>("recHitsLabelEE", "EcalRecHitsEE", true) and
+                           edm::ParameterDescription<bool>("recoverEEIsolatedChannels", false, true) and
+                           edm::ParameterDescription<bool>("recoverEEVFE", false, true) and
+                           edm::ParameterDescription<bool>("recoverEEFE", true, true) and
+                           edm::ParameterDescription<double>("EELaserMIN", 0.5, true) and
+                           edm::ParameterDescription<double>("EELaserMAX", 8.0, true)) or
+                     true >> edm::EmptyGroupDescription());
+
+    confDesc.addWithDefaultLabel(desc);
+  }
+
+  EcalRecHitProducerPortable::EcalRecHitProducerPortable(const edm::ParameterSet& ps)
+      : isPhase2_{ps.getParameter<bool>("isPhase2")},
+        uncalibRecHitsTokenEB_{consumes(ps.getParameter<edm::InputTag>("uncalibrecHitsInLabelEB"))},
+        uncalibRecHitsTokenEE_{isPhase2_ ? device::EDGetToken<InputProduct>{}
+                                         : consumes(ps.getParameter<edm::InputTag>("uncalibrecHitsInLabelEE"))},
+        recHitsTokenEB_{produces(ps.getParameter<std::string>("recHitsLabelEB"))},
+        recHitConditionsToken_{esConsumes()},
+        recHitParametersToken_{esConsumes()} {
+    if (!isPhase2_) {
+      recHitsTokenEE_ = produces(ps.getParameter<std::string>("recHitsLabelEE"));
+    }
```

Comment:
could this one be moved to the initialiser list, too ?
```c++
        recHitsTokenEE_{isPhase2_ ? device::EDPutToken<OutputProduct>{} : produces(ps.getParameter<std::string>("recHitsLabelEE"))}

Reply:
No I tried that before and it does not compile.
```
error: operands to '?:' have different types 'alpaka_serial_sync::device::EDPutToken<PortableHostCollection<EcalRecHitSoALayout<> > >' and 'alpaka_serial_sync::ProducerBaseAdaptor<alpaka_serial_sync::ProducerBase<edm::stream::EDProducer>, edm::Transition::Event>'
```

Reply:
(not for this PR)

@makortel an we add a method like `does_not_produce()` to be used like
```c++
    recHitsTokenEE_{isPhase2_ ? does_not_produce() : produces(ps.getParameter<std::string>("recHitsLabelEE"))}
```

or even `produces_if()` to be used like
```c++
    recHitsTokenEE_{produces_if(not isPhase2_, ps.getParameter<std::string>("recHitsLabelEE"))}
```

?

Reply:
--> https://github.com/cms-sw/cmssw/issues/46960